### PR TITLE
Code Refactor of IssueWatch related things

### PR DESCRIPTION
**Pull request by 6543**
taken from #9535

### CODE REVIEW

File: models/issue_watch.go [hunk 1]
```diff
@@ -97,14 +100,11 @@ func getIssueWatchers(e Engine, issueID int64, listOptions ListOptions) (watches
...
 func removeIssueWatchersByRepoID(e Engine, userID int64, repoID int64) error {
-	iw := &IssueWatch{
-		IsWatching: false,
-	}
 	_, err := e.
 		Join("INNER", "issue", "`issue`.id = `issue_watch`.issue_id AND `issue`.repo_id = ?", repoID).
 		Cols("is_watching", "updated_unix").
 		Where("`issue_watch`.user_id = ?", userID).
-		Update(iw)
+		Delete(new(IssueWatch))
```
Comment (lunny) [inline models/issue_watch.go:109]: The original code is more clear.

Comment (guillep2k) [reply models/issue_watch.go:109]: `is_watching = false` vs. Delete, I don't think it's equivalent. If I'm not mistaken, a `false` row means the user does not want to receive notifications, but a missing row means that the user may or may not receive them according to other logic (e.g. repository watch, issue participation, etc.).

Comment (6543) [reply models/issue_watch.go:109]: this function (`removeIssueWatchersByRepoID`) is used in:
 * removeAllRepositories -> to delete IssueWatches
 * removeRepository -> to delete IssueWatches
 * removeTeamMember ->  to delete IssueWatches
 * DeleteCollaboration  ->  to delete IssueWatches

so until NOW the function didn't do what a developer expect from its name!
on master-branch all the Delete/remove functions witch call removeIssueWatchersByRepoID set an explicit DontWatch instead of removing IssueWatche! so this fix it
@lunny

Comment (6543) [reply models/issue_watch.go:109]: @guillep2k as you wrote ...

File: models/issue_watch.go [hunk 2]
```diff
@@ -97,39 +100,10 @@ func getIssueWatchers(e Engine, issueID int64, listOptions ListOptions) (watches
...
 	return err
 }
-
-// LoadWatchUsers return watching users
```
Comment (6543) [inline models/issue_watch.go:113]: iw has no User to load ... so remove this function

**Comment by 6543** [comment]
@lunny -> https://github.com/go-gitea/gitea/pull/10401#discussion_r382928559

File: models/issue_watch.go [hunk 3]
```diff
@@ -97,14 +100,11 @@ func getIssueWatchers(e Engine, issueID int64, listOptions ListOptions) (watches
 }
 
 func removeIssueWatchersByRepoID(e Engine, userID int64, repoID int64) error {
-	iw := &IssueWatch{
-		IsWatching: false,
-	}
 	_, err := e.
 		Join("INNER", "issue", "`issue`.id = `issue_watch`.issue_id AND `issue`.repo_id = ?", repoID).
 		Cols("is_watching", "updated_unix").
```
Comment (guillep2k) [inline models/issue_watch.go]: If we're not updating, what do we need the `Cols()` list for?

File: models/issue_watch.go (hunk 1 above)
Comment (guillep2k) [reply models/issue_watch.go:109]: Now that you mention it, it seems that when a user is removed as collaborator from a repo, all watches he had on its issues are removed as well, even if they still had access to them (e.g. as site admin)... 🤔
Same thing happens in `(t *Team) removeRepository()` and `(t *Team) removeAllRepositories()`.

File: models/issue_watch.go (hunk 3 above)
Comment (6543) [reply models/issue_watch.go]: good point!

**Comment by 6543** [comment]
@guillep2k if it is a good idear or not ... this pull is still an impruvement in the right direction

**Comment by 6543** [comment]
@guillep2k it all started with #4201

**Review by guillep2k — APPROVED** [review]
LGTM. Just a small nit.

**Comment by guillep2k** [comment]
I think that the CI fail is related. I restarted the tests a couple of times and it was always blocked in `models.createOrUpdateIssueNotifications` (among others). Maybe there's a `e.Close()` missing?

**Comment by 6543** [comment]
> createOrUpdateIssueNotifications

@guillep2k no a Close() is here - the failure looks unrelated to me for example It already passed several times:
* https://drone.gitea.io/go-gitea/gitea/21724
* https://drone.gitea.io/go-gitea/gitea/21792

so its weared that if failed after adding your code sugestion (witch only add a space)

**Comment by 6543** [comment]
`make test-mysql` Passed (with #10446 and beside https://hastebin.com/kemixeponi.log) on my local pc

**Comment by 6543** [comment]
@lunny found the bug https://github.com/go-gitea/gitea/pull/10401/commits/19e34ffe3f5fcb35f89991b69840ccae97257bc9 do you have time to look at it again?

File: models/issue_watch.go (hunk 1 above)
Comment (lunny) [reply models/issue_watch.go:109]: @guillep2k That's a problem.

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
ready to :rocket:

File: models/issue_watch.go [hunk 4]
```diff
@@ -68,10 +68,14 @@ func getIssueWatch(e Engine, userID, issueID int64) (iw *IssueWatch, exists bool
 // but avoids joining with `user` for performance reasons
 // User permissions must be verified elsewhere if required
 func GetIssueWatchersIDs(issueID int64) ([]int64, error) {
+	return getIssueWatchersIDs(x, issueID, true)
+}
+
+func getIssueWatchersIDs(e Engine, issueID int64, watching bool) ([]int64, error) {
 	ids := make([]int64, 0, 64)
 	return ids, x.Table("issue_watch").
```
Comment (6543) [inline models/issue_watch.go]:
```suggestion
	return ids, e.Table("issue_watch").
```

**Comment by 6543** [comment]
Ping lgtm bot

**Comment by 6543** [comment]
wake up lgtm bot

File: modules/git/repo_branch.go [hunk 5]
```diff
@@ -48,6 +48,9 @@ type Branch struct {
 
 // GetHEADBranch returns corresponding branch of HEAD.
 func (repo *Repository) GetHEADBranch() (*Branch, error) {
+	if repo == nil {
```
Comment (guillep2k) [inline modules/git/repo_branch.go:51]: This unrelated change should not be here.

Comment (6543) [reply modules/git/repo_branch.go:51]: @guillep2k schould I send a revert pull for https://github.com/go-gitea/gitea/pull/10401/commits/99867f3505d9e00dd40da7b471b6b0a1382fd370 ?

Comment (6543) [reply modules/git/repo_branch.go:51]: When Tests failed a lot I add this so test dont crash only fail ...
